File: services/evidence/merkle.py
"""RFC 6962 Merkle tree over SHA-256, with domain-separated leaves and nodes.

Pure bytes in, pure bytes out. This module knows nothing about ARES types: it
takes a list of leaf preimages and returns hashes, so it can be tested in
isolation and reused by anything. It deliberately does **not** police duplicate
preimages -- a Merkle tree legitimately may contain them, and the decision that
an ARES evidence bundle may not is a policy call that lives in bundle.py.

Why RFC 6962 and not the Bitcoin construction
---------------------------------------------
Bitcoin hashes a lone right-hand child as H(x||x). That makes the 3-leaf tree
[a,b,c] and the 4-leaf tree [a,b,c,c] produce the *identical* root
(CVE-2012-2459): two different leaf multisets, one root. For an evidence anchor
that means the root attests to a finding set that was never produced.

Our leaves happen to be unique, so a duplicate check would paper over it. But
patching a malleable tree with a uniqueness invariant is the fragile design --
it breaks the moment someone relaxes the invariant. RFC 6962 splits at the
largest power of two below n and never duplicates, so the malleability is
structurally impossible rather than checked-against, the shape is canonical for
each n, and the published second-preimage reduction is stated for exactly this
construction with exactly these prefixes.

Why the prefixes
----------------
Without them, H(L||R) for an internal node is drawn from the same distribution
as H(x) for a leaf x = L||R. An attacker could present an internal node as a
leaf and produce a proof that verifies for data that was never a leaf -- an
n-leaf tree reinterpreted as a tree of fewer, larger ones. Concretely here:
take an anchored 8-finding root and produce a valid-looking inclusion proof for
a *fabricated* finding, backed by an on-chain root.

The 16-byte domain tag is a second, cheaper layer. It stops any other SHA-256
Merkle tree in the ecosystem (an SPL merkle-distributor root, a concurrent
merkle-tree node) from ever being replayed as an ARES evidence node, and it
version-gates a future v2 of the encoding.

The values here are pinned by services/evidence/vectors/merkle_vectors.json,
which is also what the Rust spec crate and the TypeScript oracle assert against
-- that shared file is the only thing proving the three implementations agree.
"""

from __future__ import annotations

import hashlib

from canonical import EvidenceError, lp

#: Exactly 16 bytes, so it needs no length prefix of its own to be unambiguous.
DOMAIN = b"ares.evidence.v1"

LEAF_PREFIX = b"\x00"
NODE_PREFIX = b"\x01"
COMMITMENT_PREFIX = b"\x02"
TARGET_BINDING_PREFIX = b"\x03"

assert len(DOMAIN) == 16, "DOMAIN is length-prefix-free only because it is fixed at 16 bytes"

#: RFC 6962's MTH({}) -- the hash of the empty string, carrying neither the leaf
#: nor the node prefix, so an empty tree's root cannot be confused with either.
#: Recognisable on sight (e3b0c442...), which is a real benefit when a third
#: party is eyeballing an anchored value.
EMPTY_ROOT = hashlib.sha256(b"").digest()

HASH_SIZE = 32


def leaf_hash(preimage: bytes) -> bytes:
    return hashlib.sha256(LEAF_PREFIX + DOMAIN + preimage).digest()


def node_hash(left: bytes, right: bytes) -> bytes:
    if len(left) != HASH_SIZE or len(right) != HASH_SIZE:
        raise EvidenceError("node_hash takes two 32-byte digests")
    return hashlib.sha256(NODE_PREFIX + DOMAIN + left + right).digest()


def largest_pow2_below(n: int) -> int:
    """RFC 6962's k: the largest power of two strictly less than n."""
    if n < 2:
        raise EvidenceError(f"k is only defined for n >= 2, got {n}")
    return 1 << (n - 1).bit_length() - 1


def _build(preimages: list[bytes], base: int, proofs: dict[int, list]) -> bytes:
    """Compute the root and, in the same pass, every leaf's audit path.

    Siblings are appended as the recursion unwinds, so each proof comes out
    bottom-up (deepest sibling first) -- the order RFC 6962's PATH() produces
    and the order verify_inclusion consumes.
    """
    n = len(preimages)
    if n == 1:
        return leaf_hash(preimages[0])
    k = largest_pow2_below(n)
    left = _build(preimages[:k], base, proofs)
    right = _build(preimages[k:], base + k, proofs)
    for i in range(base, base + k):
        proofs[i].append(("right", right))
    for i in range(base + k, base + n):
        proofs[i].append(("left", left))
    return node_hash(left, right)


def root(preimages: list[bytes]) -> bytes:
    """MTH(D[n])."""
    if not preimages:
        return EMPTY_ROOT
    return _build(list(preimages), 0, {i: [] for i in range(len(preimages))})


def build(preimages: list[bytes]) -> tuple[bytes, list[list[tuple[str, bytes]]]]:
    """Return (root, per-leaf audit paths) in one pass."""
    if not preimages:
        return EMPTY_ROOT, []
    proofs: dict[int, list] = {i: [] for i in range(len(preimages))}
    r = _build(list(preimages), 0, proofs)
    return r, [proofs[i] for i in range(len(preimages))]


def derive_positions(index: int, leaf_count: int) -> list[str]:
    """Which side each sibling sits on, derived from the tree shape alone.

    This exists so no verifier ever has to trust the positions a prover supplied.
    A position array is an attacker-controlled degree of freedom, and RFC 6962's
    unbalanced shape makes index-to-position derivation the error-prone part --
    which is exactly the part that must not be delegated to the prover. Both
    this module's verifier and the on-chain program re-derive and compare.
    """
    if not 0 <= index < leaf_count:
        raise EvidenceError(f"leaf index {index} out of range for {leaf_count} leaves")
    positions: list[str] = []

    def rec(n: int, m: int) -> None:
        if n == 1:
            return
        k = largest_pow2_below(n)
        if m < k:
            rec(k, m)
            positions.append("right")
        else:
            rec(n - k, m - k)
            positions.append("left")

    rec(leaf_count, index)
    return positions


def verify_inclusion(
    preimage: bytes,
    index: int,
    leaf_count: int,
    proof: list[tuple[str, bytes]],
    expected_root: bytes,
) -> bool:
    """Verify one leaf against a root, re-deriving the sibling sides.

    What re-deriving positions does and does not buy, stated precisely, because
    the tempting overclaim is wrong:

    It rejects a proof whose sibling *sides* disagree with the shape implied by
    `(index, leaf_count)`. That is a real constraint -- for index 4 the audit
    path is ["left"] at n=5 but ["right", "left"] at n=6, so a substituted count
    is caught.

    It does **not** make every wrong `leaf_count` detectable. For index 0 the
    path is ["right", "right", "right"] at both n=5 and n=6, so a proof built at
    n=5 still verifies if the prover claims n=6. That is inherent to any bare
    inclusion check: given the sibling digests, the root recomputes regardless of
    what size is asserted alongside them.

    The protection that actually closes this is not here. It is that
    `commitment()` binds `leaf_count` *together with* the root, so a verifier
    compares the pair against what was anchored on chain rather than trusting
    either alone -- and that RFC 6962 gives the same leaves a different root at a
    different n, so a prover cannot pick the count freely. A caller that verifies
    a proof without also checking `leaf_count` against the anchored value has
    skipped the step that matters.
    """
    expected_positions = derive_positions(index, leaf_count)
    if [side for side, _ in proof] != expected_positions:
        return False
    acc = leaf_hash(preimage)
    for side, sibling in proof:
        if len(sibling) != HASH_SIZE:
            return False
        acc = node_hash(sibling, acc) if side == "left" else node_hash(acc, sibling)
    return acc == expected_root


def target_binding(
    *,
    target_name: str,
    commit_hash: str | None,
    ares_version: str,
    operator_program_id: str | None,
    report_kind: str,
) -> bytes:
    """Bind an anchor to a specific target, so it cannot be replayed elsewhere.

    `ares_version` is technically redundant with the report digest that the
    commitment also covers, but that digest is opaque; the commitment preimage is
    published in the bundle, so this is the one place a reader can see which
    engine made the claim without holding the report.

    `report_kind` prevents a `confirmed` bundle being presented as a `scan`
    bundle, or the reverse.
    """
    return hashlib.sha256(
        TARGET_BINDING_PREFIX
        + DOMAIN
        + lp(target_name.encode("utf-8"))
        + lp((commit_hash or "").encode("utf-8"))
        + lp(ares_version.encode("utf-8"))
        + lp((operator_program_id or "").encode("utf-8"))
        + lp(report_kind.encode("utf-8"))
    ).digest()


def commitment(
    *,
    leaf_count: int,
    merkle_root: bytes,
    report_sha256: bytes,
    binding: bytes,
) -> bytes:
    """The 32 bytes an anchor actually records.

    Why a commitment and not the bare root: RFC 6962's empty tree has one
    constant root for every target on earth, and 178 of the 636 reports in the
    local corpus produce no leaves at all. Anchoring the bare empty root would
    publish "nothing found" in a form anyone could replay against any program.
    Folding in the report digest and the target binding keeps even the empty case
    unique per (target, commit, artifact bytes).

    (194 reports have `findings == []`; the smaller 178 is the number with no
    leaves, because suppressed findings are leaves too. The distinction matters
    here and is easy to conflate.)

    `leaf_count` is included so that no residual shape ambiguity can change what
    was committed, independently of the tree construction.
    """
    for name, value in (("merkle_root", merkle_root), ("report_sha256", report_sha256), ("binding", binding)):
        if len(value) != HASH_SIZE:
            raise EvidenceError(f"{name} must be 32 bytes, got {len(value)}")
    if not 0 <= leaf_count <= 2**32 - 1:
        raise EvidenceError(f"leaf_count out of u32 range: {leaf_count}")
    return hashlib.sha256(
        COMMITMENT_PREFIX
        + DOMAIN
        + leaf_count.to_bytes(4, "big")
        + merkle_root
        + report_sha256
        + binding
    ).digest()

File: services/evidence/onchain/spec/src/merkle.rs
//! RFC 6962 Merkle constants and proof verification, generic over the hasher.
//!
//! Generic over `H` so this module stays dependency-free: the program supplies
//! the `sol_sha256` syscall, the host tests supply `sha2`. One implementation,
//! two hashers, no way for them to disagree about the construction.
//!
//! These constants must match `services/evidence/merkle.py` exactly, and both
//! are pinned by `services/evidence/vectors/merkle_vectors.json`.

/// Domain tag. Exactly 16 bytes, so it needs no length prefix of its own.
///
/// Bumping this is how a v2 leaf encoding is version-gated: it changes every
/// hash in the scheme, so a v1 proof can never be replayed against a v2 root.
pub const DOMAIN: &[u8; 16] = b"ares.evidence.v1";

/// Prefix for a leaf hash.
pub const LEAF_PREFIX: u8 = 0x00;
/// Prefix for an internal node hash.
pub const NODE_PREFIX: u8 = 0x01;
/// Prefix for the anchored commitment.
pub const COMMITMENT_PREFIX: u8 = 0x02;
/// Prefix for the target binding.
pub const TARGET_BINDING_PREFIX: u8 = 0x03;

/// Length of a SHA-256 digest.
pub const HASH_SIZE: usize = 32;

/// Which side a proof sibling sits on.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Side {
    /// Sibling is the left child; the accumulator is the right.
    Left,
    /// Sibling is the right child; the accumulator is the left.
    Right,
}

/// `MTH({})` -- SHA-256 of the empty string.
///
/// Carries neither the leaf nor the node prefix, so an empty tree's root cannot
/// be confused with either. A clean audit is a legitimate thing to anchor, and
/// 178 of the 636 reports in the local corpus produce no leaves at all, so this
/// is the common case rather than an edge case.
pub const EMPTY_ROOT: [u8; HASH_SIZE] = [
    0xe3, 0xb0, 0xc4, 0x42, 0x98, 0xfc, 0x1c, 0x14, 0x9a, 0xfb, 0xf4, 0xc8, 0x99, 0x6f, 0xb9, 0x24,
    0x27, 0xae, 0x41, 0xe4, 0x64, 0x9b, 0x93, 0x4c, 0xa4, 0x95, 0x99, 0x1b, 0x78, 0x52, 0xb8, 0x55,
];

/// Preimage of a leaf hash: `0x00 || DOMAIN || payload`.
pub fn leaf_preimage(payload: &[u8]) -> Vec<u8> {
    let mut out = Vec::with_capacity(1 + DOMAIN.len() + payload.len());
    out.push(LEAF_PREFIX);
    out.extend_from_slice(DOMAIN);
    out.extend_from_slice(payload);
    out
}

/// Preimage of an internal node hash: `0x01 || DOMAIN || left || right`.
///
/// The prefixes are what give second-preimage resistance. Without them,
/// `H(L||R)` for a node is drawn from the same distribution as `H(x)` for a leaf
/// `x = L||R`, so an internal node could be presented as a leaf and used to
/// prove inclusion of data that was never a leaf -- a fabricated finding backed
/// by a real anchored root.
pub fn node_preimage(left: &[u8; HASH_SIZE], right: &[u8; HASH_SIZE]) -> [u8; 81] {
    let mut out = [0u8; 81];
    out[0] = NODE_PREFIX;
    out[1..17].copy_from_slice(DOMAIN);
    out[17..49].copy_from_slice(left);
    out[49..81].copy_from_slice(right);
    out
}

/// RFC 6962's `k`: the largest power of two strictly less than `n`.
///
/// Returns `None` for `n < 2`, where `k` is undefined.
pub fn largest_pow2_below(n: u32) -> Option<u32> {
    if n < 2 {
        return None;
    }
    Some(1u32 << (u32::BITS - 1 - (n - 1).leading_zeros()))
}

/// The sibling sides on the audit path for `index`, derived from the shape alone.
///
/// Exists so no verifier has to trust the positions a prover supplied. A
/// position array is an attacker-controlled degree of freedom, and RFC 6962's
/// unbalanced shape makes index-to-side derivation the error-prone part -- which
/// is exactly the part that must not be delegated to the prover.
///
/// Returns `None` if `index` is out of range for `leaf_count`.
pub fn derive_sides(index: u32, leaf_count: u32) -> Option<Vec<Side>> {
    if index >= leaf_count {
        return None;
    }
    let mut sides = Vec::new();
    let mut n = leaf_count;
    let mut m = index;
    // Descend, recording the side at each split, then reverse: the audit path is
    // ordered deepest-sibling-first.
    while n > 1 {
        let k = largest_pow2_below(n)?;
        if m < k {
            sides.push(Side::Right);
            n = k;
        } else {
            sides.push(Side::Left);
            m -= k;
            n -= k;
        }
    }
    sides.reverse();
    Some(sides)
}

/// Verify an inclusion proof, re-deriving the sibling sides rather than trusting
/// the ones supplied.
///
/// `hasher` is the SHA-256 implementation: the syscall on chain, `sha2` in tests.
///
/// # What `leaf_count` does and does not buy
///
/// It rejects a proof whose sibling sides disagree with the shape implied by
/// `(index, leaf_count)` -- real, since for index 4 the path is `[Left]` at n=5
/// but `[Right, Left]` at n=6. It does **not** make every wrong count
/// detectable: for index 0 the path is `[Right; 3]` at both n=5 and n=6. That is
/// inherent to any bare inclusion check, because the root recomputes from
/// whatever siblings were supplied.
///
/// What closes it is [`commitment::preimage`](crate::commitment::preimage)
/// binding `leaf_count` together with the root, so a verifier compares the pair
/// against the anchored value instead of trusting either alone.
pub fn verify_inclusion<H>(
    leaf_payload: &[u8],
    index: u32,
    leaf_count: u32,
    proof: &[([u8; HASH_SIZE], Side)],
    expected_root: &[u8; HASH_SIZE],
    hasher: H,
) -> bool
where
    H: Fn(&[u8]) -> [u8; HASH_SIZE],
{
    let Some(expected_sides) = derive_sides(index, leaf_count) else {
        return false;
    };
    if proof.len() != expected_sides.len() {
        return false;
    }
    if proof.iter().map(|(_, s)| *s).ne(expected_sides) {
        return false;
    }

    let mut acc = hasher(&leaf_preimage(leaf_payload));
    for (sibling, side) in proof {
        acc = match side {
            Side::Left => hasher(&node_preimage(sibling, &acc)),
            Side::Right => hasher(&node_preimage(&acc, sibling)),
        };
    }
    &acc == expected_root
}
